Extensions/AuthPropertiesTokenCacheHelper.cs

```diff
@@ -0,0 +1,100 @@
+﻿using System;
+using System.Security.Claims;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Authentication.Cookies;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Identity.Client;
+
+namespace TodoListService.Extensions
+{
+    public class AuthPropertiesTokenCacheHelper
+    {
+        private const string TokenCacheKey = ".UserTokenCache";
+
+        private HttpContext _httpContext;
+        private ClaimsPrincipal _principal;
+        private AuthenticationProperties _authProperties;
+        private string _signInScheme;
+        public TokenCache TokenCache { get; private set; }
+
+        private AuthPropertiesTokenCacheHelper(AuthenticationProperties authProperties) : base()
+        {
+            _authProperties = authProperties;
+            TokenCache = new TokenCache();
+            TokenCache.SetBeforeAccess(BeforeAccessNotificationWithProperties);
+            TokenCache.SetAfterAccess(AfterAccessNotificationWithProperties);
+            TokenCache.SetBeforeWrite(BeforeWriteNotification);
+        }
+
+        private AuthPropertiesTokenCacheHelper(HttpContext httpContext, string signInScheme) : base()
+        {
+            _httpContext = httpContext;
+            _signInScheme = signInScheme;
+            TokenCache = new TokenCache();
+            TokenCache.SetBeforeAccess(BeforeAccessNotificationWithContext);
+            TokenCache.SetAfterAccess(AfterAccessNotificationWithContext);
+            TokenCache.SetBeforeWrite(BeforeWriteNotification);
+        }
+
+        public static TokenCache ForCodeRedemption(AuthenticationProperties authProperties)
+        {
```

Comment:
Null check on parameters

Reply:
Is this necessary if the constructors are private and only used within this class?

Reply:
probably still a good idea for the factory method here to not accept null (though in general all this null param checking is too much. should have been possible to make that declaratively.

---
In reply to: [215816712](https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/pull/7#discussion_r215816712) [](ancestors = 215816712)